replica/table.cc

```diff
@@ -3496,6 +3496,63 @@ future<row_locker::lock_holder> table::push_view_replica_updates(shared_ptr<db::
+// This function returns true if the read-before-write is not needed.
+// It returns false if for any reason, to correctly handle any of the views
+// is is necessary to do a read before write.
+// The list of views can be assumed to be non-empty (in the no-views case,
+// this function won't be called).
+static bool may_skip_read_before_write(
+        const std::vector<db::view::view_and_base>& views,
+        const mutation& m) {
+    // If *any* of the views has a key column which isn't a base-table key
+    // column, we need to know the previous value of that column to perform
+    // the view update (which may require deleting an old row), so a read-
+    // before-write *is* necessary - i.e., we need to return "false":
+    if (std::ranges::any_of(views, [] (const auto& v) {
+            return v.view->view_info()->has_base_non_pk_columns_in_view_pk(); })) {
+        return false;
+    }
+    // If we're still here, it means all the views have the same key columns
+    // as the base table (possibly in a different order).
+    // But if the base mutation affects more than a single row - i.e. deleting
+    // a whole partition or a clustering range - then because the key columns
+    // might be in different order in the view key, it might not be possible
+    // to convert the single base tombstone into a single view tombstone,
+    // and we *do* need to perform the read-before write to get the list of
+    // rows and produce tombstones for them individually.
+    // TODO: With more effort, the read-before-write can still be avoided if
+    // the key column reordering is not "too much". One example in the case of
+    // a partition tombstone if the view's partition key has the same columns
+    // as the base's (is_partition_key_permutation_of_base_partition_key).
+    // Another example when a clustering prefix being deleted in the base also
+    // happens to be a contiguous clustering prefix in the view.
+    if (m.partition().partition_tombstone() || !m.partition().row_tombstones().empty()) {
+        return false;
+    }
+    // We have a separate code path for secondary-index of a collection column
+    // which currently incorrectly relies on read-before-write in the deletion
+    // case: It wants to optimize the case of deletion of collection item that
+    // didn't previously exist and checks the "existing" (read-before-write
+    // result) to do that. So before this bug is fixed, we need to turn of the
+    // read-before-write avoidance if any of the views indexes a collection.
+    // FIXME: Fix the collection indexing bugs and remove this if().
+    if (std::ranges::any_of(views, [] (const auto& v) {
+            return std::ranges::any_of(v.view->partition_key_columns(), [] (const column_definition& cdef) {
+                return cdef.is_computed() && nullptr != dynamic_cast<const collection_column_computation*>(&cdef.get_computation());
+            });})) {
+        return false;
+    }
```

Comment:
We should perform this calculation ahead of time, just like we perform similar calculations in prepared statements. Otherwise this imposes a (small) slowdown on every view update.

Perhaps we could store the result in the view schema.

Note the dynamic_cast can be especially slow.

Reply:
Ah, it should be stored in the table schema. Do we update the table schema when we add a view?

Reply:
You're right, this would be ideal although the number of these flags we have already makes it kind of ugly, I'll look into doing this.
Please note that:
1. I believed the outer loop (any_of) will still be necessary, because the decision of to not do read-before-write depends on all the views involved in this write, which I believed is not necessarily all the views of the table and may depend on filtering and selected columns. I'm no longer sure this is correct - I'll need to review the code again.
2. The order of the conditions in the function may_skip_read_before_write() is deliberate - it returns false (read-before-write IS needed) as quickly as possible. It only reaches the slowest condition - the dynamic_cast check (which also shouldn't be very slow, by the way!) - when the answer will be "true" - in which case the cost of the checks is negligible compared to the read-before-write that we now avoided.

But I'll give all these issues a second look.